File: codewars/Fun with tree - max sum/Fun with trees - max sum.js
function maxSum(node) {
    if (!node) return 0;
    return node.value + Math.max(maxSum(node.left), maxSum(node.right));
}
const tree1 = {
    value: 17,
    left: {
        value: 3,
        left: { value: 2, left: null, right: null },
        right: null
    },
    right: {
        value: -10,
        left: { value: 16, left: null, right: null },
        right: {
            value: 1,
            left: { value: 13, left: null, right: null },
            right: null
        }
    }
};
console.log(maxSum(tree1));

const tree2 = {
    value: 5,
    left: {
        value: 4,
        left: { value: -80, left: null, right: null },
        right: { value: -60, left: null, right: null }
    },
    right: {
        value: 10,
        left: { value: -90, left: null, right: null },
        right: null
    }
};
console.log(maxSum(tree2));

File: codewars/Fun with tree - max sum/Fun_with_Trees_Max_Sum.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    public:
    static int maxSum(TreeNode* root) {
        if (!root) return root -> value;
        if (!root -> left) return root -> value + maxSum(root -> right);
        if (!root -> right) return root -> value + maxSum(root -> left);
        return root -> value + max(maxSum(root -> left), maxSum(root -> right));
    }
};
